Photo-Editor/Photo-Editor/Content/MainButton.swift

```diff
@@ -33,34 +33,139 @@ struct MainButton: View {
                 }
 
                 NavigationLink(
-                    destination: EditImageView(image: selectedImage ?? UIImage()),
+                    destination: EditImageView(
+                        image: selectedImage ?? UIImage(),
+                        onSave: { savedImageOrUrl in
+                            print("onSave closure called in NavigationLink.")
+
+                            Task {
+                                if let savedImage = savedImageOrUrl as? UIImage {
+                                    print("UIImage received, starting Task to save...")
+                                    await saveImageAndAddToLibrary(image: savedImage)
+                                    print("Finished Task to save image.")
+                                } else if let imageUrlString = savedImageOrUrl as? String {
+                                    print("Cloudinary URL received: \(imageUrlString)")
+                                    if let downloadedImage = await downloadImage(from: imageUrlString) {
+                                        await saveImageAndAddToLibrary(image: downloadedImage)
+                                        print("Image downloaded and saved to photo library.")
+                                    } else {
+                                        print("Failed to download image from Cloudinary URL.")
+                                    }
+                                } else {
+                                    print("Unknown type received from onSave closure.")
+                                }
+                            }
+                        }
+                    ),
                     isActive: $isEditImageViewActive
                 ) {
                     EmptyView()
                 }
                 .hidden()
 
+                .hidden()
+
                 .onChange(of: selectedImage) { newImage in
                     if let newImage = newImage {
                         print("New image selected")
-                        
-                        if recentImages.count < 15 {
-                            recentImages.append(newImage)
-                        } else {
-                            recentImages.removeFirst()
-                            recentImages.append(newImage)
-                        }
                         isEditImageViewActive = true
-                        print("Image added to recentImages, total count: \(recentImages.count)")
                     } else {
                         print("No image selected")
                     }
                 }
             }
         }
     }
+
+    func downloadImage(from urlString: String) async -> UIImage? {
```

Comment:
These functions should be added to the Controller class mentioned in the previous comment.